File: python/features/steps/issue_show_steps.py
"""Behave steps for issue display."""

from __future__ import annotations

import os
from datetime import datetime, timezone

import click
from behave import given, then, when

from features.steps.shared import (
    build_issue,
    load_project_directory,
    read_issue_file,
    run_cli,
    write_issue_file,
)
from kanbus.config_loader import load_project_configuration
from kanbus.issue_display import format_issue_for_display
from kanbus.models import DependencyLink, IssueComment


@given('an issue "{identifier}" exists with title "{title}"')
def given_issue_exists_with_title_generic(
    context: object, identifier: str, title: str
) -> None:
    project_dir = load_project_directory(context)
    issue = build_issue(identifier, title, "task", "open", None, [])
    write_issue_file(project_dir, issue)


@given('issue "kanbus-aaa" has status "open" and type "task"')
def given_issue_status_and_type(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, "kanbus-aaa")
    issue = issue.model_copy(update={"status": "open", "issue_type": "task"})
    write_issue_file(project_dir, issue)


@when('I run "kanbus show kanbus-aaa"')
def when_run_show(context: object) -> None:
    run_cli(context, "kanbus show kanbus-aaa")


@when('I run "kanbus show kanbus-desc"')
def when_run_show_desc(context: object) -> None:
    run_cli(context, "kanbus show kanbus-desc")


@when('I run "kanbus show kanbus-aaa --json"')
def when_run_show_json(context: object) -> None:
    run_cli(context, "kanbus show kanbus-aaa --json")


@when('I run "kanbus show kanbus-labels"')
def when_run_show_labels(context: object) -> None:
    run_cli(context, "kanbus show kanbus-labels")


@when('I run "kanbus show kanbus-missing"')
def when_run_show_missing(context: object) -> None:
    run_cli(context, "kanbus show kanbus-missing")


@when('I format issue "kanbus-labels" for display')
def when_format_issue_for_display(context: object) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, "kanbus-labels")
    context.formatted_output = format_issue_for_display(issue)


@when('I format issue "{identifier}" for display')
def when_format_issue_for_display_generic(context: object, identifier: str) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    context.formatted_output = format_issue_for_display(issue)


@when('I format issue "{identifier}" for display with color enabled')
def when_format_issue_for_display_with_color(context: object, identifier: str) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    config_path = project_dir.parent / ".kanbus.yml"
    configuration = (
        load_project_configuration(config_path) if config_path.exists() else None
    )
    command = click.Command("test")
    click_context = click.Context(command, color=True)
    with click_context.scope():
        context.formatted_output = format_issue_for_display(
            issue, configuration=configuration
        )


@when(
    'I format issue "{identifier}" for display with color enabled without configuration'
)
def when_format_issue_display_without_configuration(
    context: object, identifier: str
) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    command = click.Command("test")
    click_context = click.Context(command, color=True)
    with click_context.scope():
        context.formatted_output = format_issue_for_display(issue, configuration=None)


@when('I format issue "{identifier}" for display with NO_COLOR set')
def when_format_issue_display_no_color(context: object, identifier: str) -> None:
    if not hasattr(context, "original_no_color"):
        context.original_no_color = os.environ.get("NO_COLOR")
    os.environ["NO_COLOR"] = "1"
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    context.formatted_output = format_issue_for_display(issue, configuration=None)


@then("the formatted output should contain ANSI color codes")
def then_formatted_output_contains_ansi(context: object) -> None:
    output = getattr(context, "formatted_output", "")
    assert "\x1b[" in output


@then('the formatted output should contain text "{text}"')
def then_formatted_output_contains_text(context: object, text: str) -> None:
    output = getattr(context, "formatted_output", "")
    assert text in output


@given('issue "{identifier}" has dependency "{target}" of type "{dependency_type}"')
def given_issue_has_dependency(
    context: object, identifier: str, target: str, dependency_type: str
) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    dependency = DependencyLink(target=target, type=dependency_type)
    issue = issue.model_copy(update={"dependencies": [dependency]})
    write_issue_file(project_dir, issue)


@given(
    'issue "{identifier}" has a comment from "{author}" with text "{text}" and no id'
)
def given_issue_comment_no_id(
    context: object, identifier: str, author: str, text: str
) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    comment = IssueComment(
        id=None,
        author=author,
        text=text,
        created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
    )
    issue = issue.model_copy(update={"comments": [comment]})
    write_issue_file(project_dir, issue)


@given(
    'issue "{identifier}" has a comment from "{author}" with text "{text}" and id "{comment_id}"'
)
def given_issue_comment_with_id(
    context: object, identifier: str, author: str, text: str, comment_id: str
) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    comment = IssueComment(
        id=comment_id,
        author=author,
        text=text,
        created_at=datetime(2026, 2, 11, tzinfo=timezone.utc),
    )
    issue = issue.model_copy(update={"comments": [comment]})
    write_issue_file(project_dir, issue)

File: rust/features/steps/issue_show_steps.rs
use std::fs;
use std::path::PathBuf;

use cucumber::{given, then, when};

use kanbus::config_loader::load_project_configuration;
use kanbus::file_io::load_project_directory;
use kanbus::issue_display::format_issue_for_display;
use kanbus::models::IssueData;

use crate::step_definitions::initialization_steps::KanbusWorld;

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

#[given("issue \"kanbus-aaa\" has status \"open\" and type \"task\"")]
fn given_issue_status_type(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let issue_path = project_dir.join("issues").join("kanbus-aaa.json");
    let contents = fs::read_to_string(&issue_path).expect("read issue");
    let mut payload: serde_json::Value = serde_json::from_str(&contents).expect("parse");
    payload["status"] = "open".into();
    payload["type"] = "task".into();
    let updated = serde_json::to_string_pretty(&payload).expect("serialize");
    fs::write(&issue_path, updated).expect("write issue");
}

#[when(expr = "I format issue {string} for display")]
fn when_format_issue_display_generic(world: &mut KanbusWorld, identifier: String) {
    let project_dir = load_project_dir(world);
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(&issue_path).expect("read issue");
    let issue: IssueData = serde_json::from_str(&contents).expect("parse issue");
    let config_path = project_dir
        .parent()
        .unwrap_or(&project_dir)
        .join(".kanbus.yml");
    let configuration = if config_path.exists() {
        Some(load_project_configuration(&config_path).expect("load configuration"))
    } else {
        None
    };
    world.formatted_output = Some(format_issue_for_display(
        &issue,
        configuration.as_ref(),
        false,
        false,
    ));
}

#[when(expr = "I format issue {string} for display with NO_COLOR set")]
fn when_format_issue_display_no_color(world: &mut KanbusWorld, identifier: String) {
    std::env::set_var("NO_COLOR", "1");
    when_format_issue_display_generic(world, identifier);
    std::env::remove_var("NO_COLOR");
}

#[when(expr = "I format issue {string} for display with color enabled")]
fn when_format_issue_display_with_color(world: &mut KanbusWorld, identifier: String) {
    let project_dir = load_project_dir(world);
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(&issue_path).expect("read issue");
    let issue: IssueData = serde_json::from_str(&contents).expect("parse issue");
    let config_path = project_dir
        .parent()
        .unwrap_or(&project_dir)
        .join(".kanbus.yml");
    let configuration = if config_path.exists() {
        Some(load_project_configuration(&config_path).expect("load configuration"))
    } else {
        None
    };
    world.formatted_output = Some(format_issue_for_display(
        &issue,
        configuration.as_ref(),
        true,
        false,
    ));
}

#[when(expr = "I format issue {string} for display with color enabled without configuration")]
fn when_format_issue_display_without_configuration(world: &mut KanbusWorld, identifier: String) {
    let project_dir = load_project_dir(world);
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(&issue_path).expect("read issue");
    let issue: IssueData = serde_json::from_str(&contents).expect("parse issue");
    world.formatted_output = Some(format_issue_for_display(&issue, None, true, false));
}

#[then("the formatted output should contain ANSI color codes")]
fn then_formatted_output_contains_ansi(world: &mut KanbusWorld) {
    let output = world.formatted_output.as_deref().unwrap_or("");
    assert!(output.contains("\u{1b}["));
}

#[then(expr = "the formatted output should contain text {string}")]
fn then_formatted_output_contains_text(world: &mut KanbusWorld, text: String) {
    let output = world.formatted_output.as_deref().unwrap_or("");
    assert!(output.contains(&text));
}
